google/ggrc-core
# GGRC-5031 Optimize permissions query

# Issue description

The loading of user permissions from ACL table took too long.
Two main causes behind that were:
 - slow query because the correct index was not used.
 - slow python code that translated the query results into the permissions dict

# Steps to test the changes

With the latest ggrc-qa db, log in as globcreator and check the speed of responses for any request that requires a full permissions load.

Also pay attention to the object counts and make sure that the user does not lose any access to objects that they previously had.

# Solution description

I have created an index on access control list table that can be used for group_by and added it as a hint to the query.

Then for the slow python code, we have just removed the redundant Relationship objects that were loaded. (we never check permissions on those and those are in the db just for easier propagation)


## Benchmarks (on local machine):
on ggrc-qa db for user "globcreator" with id 402:

generating permissions dict after query executed:
 - with relationships: 0.76969
 - without relationships: 0.55567

Mysql query:
original: 52019 rows in set (1.40 sec)
with index: 52019 rows in set (0.11 sec)
with index without relationships: 34007 rows in set (0.09 sec)

## Overall improvement 

before: ~ 2.2s
after: ~ 0.6s


# Sanity checklist




# PR Review checklist
